### PR TITLE
DO NOT MERGE - for internal review only  - Dag tester

### PR DESCRIPTION
**Pull request by sv3ndk**
### Overview

This PR adds an end-to-end dag tester to Airflow, located in `tests_dags/dag_tester.py`, as well as some DAG tests located in `tests_dags/tests/bash_operator_dag_tests.py`

These tests require a connection to a concurrent db (i.e. not SQLite) in the `sql_alchemy_conn` field of your `unittests.cfg`. The tests will be skipped otherwise.
### How to use
- The test author creates one or several DAG files and store them in `tests_dags/dags`. 
- The test author provides one or several units test for the dag(s), (cf examples in `tests_dags/tests/bash_operator_dag_tests.py`). This test should provide: 
  - `get_dag_file_names` to specify which dag files are involved in this test
  - `get_context` (optional): a dictionary of `Variable` to be provided to the DAG at test time
  - if we want to test the dags with a `BackfillJob`, inherit from `EndToEndBackfillJobTest` and  implement `get_backfill_params` to specify the constructor parameters of `BackfillJob`. 
  - if we want to test the dags with a `SchedulerJob`, inherit from `EndToEndSchedulerJobTest` and  implement `get_schedulerjob_params` to specify the constructor parameters of `SchedulerJob`
  - implement `post_check` with a set of post-execution validation steps 
- The DAG created by the test author may access the name of their runtime temporary folder with `Variable.get("unit_test_tmp_dir")` as well as any other context parameter provided by the test. For example, if the test implements this:
  
  ```
  def get_context(self):
      return {"country": "India",
              "product_sample_size": 10000}
  ```
  
  then the DAG will be able to access those values at test time via: 
  
  ```
  Variable.get(key="country", deserialize_json=True)
  Variable.get(key="product_sample_size", deserialize_json=True)
  ```
- The design of the test should keep in mind that the executors are ultimately starting child processes via some `airflow run blah` commands. This is particularly important to design the interactions between the test initialization, its running, and the validation of post-conditions, since they do not all run within the same python process. 
### How it works

The DAG Tester will:
- create a temporary folder for each test and copy all the tested DAGs in a subfolder `dags` of that folder
- copy an airflow configuration file to that temporary folder, adapts its `dags_folder` value according to the above. 
- write in `Variables` the location of the temporary folder in `unit_test_tmp_dir` as well as all the values provided by the test in `get_context`. (If any `Variables` would have been overwritten in the process, their values are saved and restored during `cleanup`)
- depending on what is implemented by the test, build a `BackfillJob` and/or a `SchedulerJob` with the tested dags and run it
- after each run, execute the test `post_check`
### Future possible improvements
- we currently only run the test with a `BackfillJob` with a `SequencialExecutor` and/or with a `SchedulerJob` with a `LocalExcecutor`. We could add more executors and combinations
- each test is executed with its own config file, though it's almost the same for every test => we could extend the framework to validate different kinds of configurations per test
- all tests provided so far are validating the `BashOperator`, though adding more kinds of operator should be straightforward. 
- we could extend it to run the job several times, maybe thus allowing to validate restartability of a DAG (e.g. a whole DAG that fails twice before succeeding), idempotency (a DAG that is executed several times with the same `execution_date`, with a `clear` between each run, should not produce corrupted output) or simulating addition/removal/update of DAG files between successive executions.
- we could try to extend to execution of several jobs in parallel, thus improving the validation of race condition (although test of parallel code is always tricky) 
### Limitations
- The `num_runs` parameter for `SchedulerJob` tests is artificially high in some cases, we do not always understand why more `DagRun`s than the theoretical number are needed. This has an impact on the execution time.
- The execution time of the 7 DAG tests with both kinds of executors over a period of up to 10 days requires about 20 minutes to execute on Travis => we might need to discuss when we want execute them.

### CODE REVIEW

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -361,7 +361,7 @@ def serve_logs(filename):
 def worker(args):
     # Worker to serve static log files through this simple flask app
     env = os.environ.copy()
-    env['AIRFLOW_HOME'] = settings.AIRFLOW_HOME
```
Comment (thoralf-gutierrez) [inline airflow/bin/cli.py]: Tiens, c'est louche ca, pourquoi `get_dags_folder()` pour retrouver `AIRFLOW_HOME` ?

File: airflow/models.py [hunk 2]
```diff
@@ -116,17 +114,24 @@ def __init__(
             self,
             dag_folder=None,
             executor=DEFAULT_EXECUTOR,
-            include_examples=configuration.getboolean('core', 'LOAD_EXAMPLES'),
```
Comment (thoralf-gutierrez) [inline airflow/models.py]: Should we replace this with a new accessor method ? (pas sur de pourquoi on le replace by None)

Comment (thoralf-gutierrez) [inline airflow/models.py]: Ah got it.

File: airflow/models.py [hunk 3]
```diff
@@ -116,17 +114,24 @@ def __init__(
...
+
         self.dag_folder = dag_folder
         self.dags = {}
         self.sync_to_db = sync_to_db
         self.file_last_changed = {}
         self.executor = executor
         self.import_errors = {}
+
+        # reading the config here and not as constructor arg since config can
+        # now be reloaded
```
Comment (thoralf-gutierrez) [inline airflow/models.py:132]: Et si on le mettait dans les arguments du constructeur, ce serait figé pour toujours ?

Comment (thoralf-gutierrez) [inline airflow/models.py:132]: Indeed, confirmed by :

```
>>> def compute_default(a):
...     return a*2
...
>>> class A():
...     def __init__(self,b=compute_default(2)):
...         self.b = b
...
>>> a1 = A()
>>> a1.b
4
>>> def compute_default(a):
...     return a*3
...
>>> a2 = A()
>>> a2.b
4
>>>
```

File: airflow/bin/cli.py (hunk 1 above)
Comment (svendx4f) [inline airflow/bin/cli.py]: outch, bien vu, c'est un bug, je fixe ça...

File: airflow/models.py (hunk 3 above)
Comment (sv3ndk) [inline airflow/models.py:132]: Yes, exactement, on veut que les `configuration.git_bla` soient appelés à chaque exécution du constructeur, et pas lorsque la class est crée (en fait, pour crée une class definition, python "exécute" le statement `class` => il exécute le `compute_default()` à ce moment là aussi, il ne n'exécute pas à chaque appel de la méthode

File: docs/conf.py [hunk 4]
```diff
@@ -17,12 +17,14 @@
...
 # behavior of the utils.apply_default that was hiding function headers
+import configuration as configuration
+
 os.environ['BUILDING_AIRFLOW_DOCS'] = 'TRUE'
-from airflow import settings
 
 # If extensions (or modules to document with autodoc) are in another directory,
 # add these directories to sys.path here. If the directory is relative to the
 # documentation root, use os.path.abspath to make it absolute, like shown here.
+sys.path.insert(0, configuration.get_dags_folder() + "/src/airflow")
```
Comment (thoralf-gutierrez) [inline docs/conf.py]: Je suis pas sur de comprendre ce que ceci fait

**Comment by thoralf-gutierrez** [comment]
Pushed PR to airbnb

**Pull request closed without merging**